rohmu/notifier/http.py

```diff
@@ -0,0 +1,132 @@
+"""Copyright (c) 2022 Aiven, Helsinki, Finland. https://aiven.io/"""
+from .interface import Notifier
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum, unique
+from logging import getLogger
+from queue import SimpleQueue
+from requests.exceptions import RequestException
+from typing import Any, Dict, Optional
+
+import requests
+import threading
+
+LOG = getLogger(__name__)
+
+
+@unique
+class Operation(Enum):
+    UPLOAD = "UPLOAD"
+    DELETE = "DELETE"
+    DELETE_TREE = "DELETE_TREE"
+
+
+def _get_requests_session() -> requests.Session:
+    retry = requests.adapters.Retry(
+        total=3,
+        backoff_factor=0.5,
+        status_forcelist=[429, 500, 502, 503, 504],
+        method_whitelist={"POST"},
+    )
+    adapter = requests.adapters.HTTPAdapter(max_retries=retry)
+    request_session = requests.Session()
+    request_session.mount("http://", adapter)
+    return request_session
+
+
+class HTTPNotifier(Notifier):
+    def __init__(self, url: str) -> None:
+        self._session = _get_requests_session()
```

Comment:
This leaks a `requests.Session` every time the object is instantiated.

Reply:
Why would there be more than one instance of such an object?

Reply:
I suppose anyone using rhomu in their loop could potentially create a new notifier on each iteration (even though they technically wouldn't have to

Reply:
Okay, but if people are using things like that we most likely are leaking every transfer object, the `BaseTransfer` has not `stop` or `close` method

Reply:
Is that code relying on garbage collection to release resources?

Reply:
actually `Session might` be holding connection pools/threads too?

Reply:
well, if the `BaseTransfer` is relying on the GC, since that is used to handle cycles, this should be fine too (random doc [about cycles and the gc](https://docs.python.org/3/c-api/gcsupport.html))

Sample script:

```python
from dataclasses import dataclass, field


@dataclass
class Child:
    parent: "Parent"

    def __del__(self) -> None:
        print("Child is gone")


@dataclass
class Parent:
    child: Child = field(init=False)

    def __post_init__(self) -> None:
        child = Child(self)
        self.child = child

    def __del__(self) -> None:
        print("Parent is gone")


Parent()
```

Just to be clear, I'm by no means stating this is good design. I am, however, trying not rewrite the whole thing.

Reply:
Hummm, actually let me check something. Because we are going over a thread boundary tehre.

Reply:
Good point, we'll have to make sure `BaseTransfer` can also be closed (and it might be the one responsible for cleaning up its notifiers).